### ISSUE TITLE
ResourceNotFoundException No routes found for "/eventum/htdocs/

### ISSUE THREAD

**Issue by JimFromAustin** [eventum/eventum#754]
**3.8.0**

**Description**  
Upgraded from v 2.1.1 to 3.5.6 with no issues.  Proceeded to upgrade to 3.8.0 because nothing in 3.7.x series was verified.  Ran into this issue.

**How to reproduce**  
Upgrade from 3.5.6 to 3.8.0 and it will occur. 

Guessing this has something to do with Symphony, which I know nothing about.  I read the install instructions from  https://github.com/eventum/eventum/blob/master/docs/wiki/System-Admin/Doing-a-fresh-install.md  and under Installation Process it says following:

PLEASE NOTE: The whole eventum directory should NOT be accessible under the webserver, only /path-to-eventum/htdocs/

Can you explain what this means?  I don't understand what this means?  Could you provide an example file structure demonstrating what this means? What part of the eventum directory should not be accessible, and where should it be located?  Are you referring to using aliases?  What part of the directory structure should be aliased?  Sorry, just totally confused on how to setup.

I tried doing a fresh install, not an upgrade.  Was hoping this may give a clue, but wasn't able to get it running. 

Thanks guys, long time user of Eventum, just realized you guys had continued development from the Sun Micro days.  Thank you!

*******************************************************************
BTW:  Here is my upgrade path in case anyone cares:

Took this upgrade path based on release notes when upgrading from older version and selected releases that had been verified. Was running Apache 2.4.39, MySQL 5.7.26

    2.1.1 starting point
    2.2 : run script in browser
    3.0.12 : This will be last release supporting PHP 5.3, next version will require PHP 5.5 and be versioned as 3.1.0
    3.1.10 : had to restart script a few times until #59, #59 had 1st line in script remove_impact_analysis.sql DROP TABLE eventum_issue_requirement;
    3.2.0 : #69 gets error "Syntax error or access violation: 1171 All parts of a PRIMARY KEY must be NOT NULL; if you need NULL in a key, use UNIQUE instead" 

Alteration of table eventum_user_alias is the problem. Change script to the following: ALTER TABLE eventum_user_alias modify `ual_email` varchar(255) CHARACTER SET latin1 NOT NULL DEFAULT ;;

    3.3.4 : no issues
    3.4.4 : no issues
    3.5.6 : no issues
    3.7.4 : switched to PHP 7.1.29 because was a requirement starting in 3.6.0 

Parse error: syntax error, unexpected ':', expecting '{' in C:\wamp64\www\eventum\init.php on line 33,

**Comment by glensc** [comment]
there are existing issues on this same problem, I think the workaround is to prefix all routes with the path you use, if you don't install to virtualhost root.

**Comment by JimFromAustin** [comment]
Thank you for getting back.  My install is wamp64 on Windows 10.  Install path is c:\wamp64\www\eventum

Are you suggesting to modify routes.yml  like this:

Default:
root:
    path: /
    controller: Eventum\Controller\IndexController::defaultAction

Modify default like this:
root:
    path: c:\wamp64\www\eventum
    controller: Eventum\Controller\IndexController::defaultAction

**Comment by glensc** [comment]
the `path: ` should be path of your webroot:

`path: /eventum` or `path: /eventum/htdocs` or whatever you have alias to the eventum `htdocs` folder.

**Comment by JimFromAustin** [comment]
Thank you I will try that.  in my Windows WAMP install, the webroot is C:\wamp64\www.   I'm guessing you are developing on a Linux based system where it would be \htdocs.   Could you explain why there is also a htdocs subdirectory under the eventum directory?  (\htdocs\eventum\htdocs)   I'm assuming this is not to be confused with an Apache webserver webroot \htdocs?

**Comment by glensc** [comment]
You are still talking about filesystem paths, but I'm speaking at URL that the installation is available.

If you unpack eventum installation to `C:\wamp64\www`, then webroot will be `/htdocs`: http://example.com/htdocs.

But as the [documentation](https://github.com/eventum/eventum/blob/v3.8.5/docs/wiki/System-Admin/Doing-a-fresh-install.md#installation-process) said, you must not install to webroot, but install to an arbitrary path (say `c:/eventum`) and create [virtualhost](https://httpd.apache.org/docs/2.4/vhosts/examples.html) or use [alias](https://httpd.apache.org/docs/2.4/mod/mod_alias.html) pointing to that directory.

1. accessible from http://example.com/eventum

```apache
Alias /eventum c:/eventum/htdocs
```

2. accessible from http://example.com

```apache
<VirtualHost *:80>
  DocumentRoot c:/eventum/htdocs
</VirtualHost>
```

**Comment by glensc** [comment]
files that are publicly accessible (over http) are in `htdocs` folder, everything outside **must not be accessible** from web. that's why the subdir that you point your webserver configuration to subdir.

**Comment by JimFromAustin** [comment]
Finally getting back to this.  

#1 created alias  => Alias /test2 "C:/wamp64/apps/test2/htdocs"
#2 unzipped entire eventum 3.8.7 to alias directory and tried fresh install by pointing browser to the alias.  ( http://localhost/test2)
#3 It automatically redirects me to http://localhost/test2/setup/ and then gives following error

"No route found for "GET /test2/setup"

The last execution point where NotFoundHttpException is thrown is:
 in C:\wamp64\apps\test2\vendor\symfony\http-kernel\EventListener\RouterListener.php (line 136)

Any idea what I may be doing wrong?

**Comment by glensc** [comment]
As I told, I believe you need to edit `routes.yml` to prefix your alias too all routes if your Eventum setup is not in root `/` but at alias `/test2`

**Comment by JimFromAustin** [comment]
I followed the installation instructions as you suggested.  Eventum was installed in aliased directory, not in root.  That means I should not have to edit routes.yml  Is this correct?

Here is what I did:

 My root is C:/wamp64/www  - I did not put any eventum files in this directory

#1) The directory I unzipped eventum in is: C:/wamp64/apps/test2/htdocs
#2) I made 'test2' the alias for C:/wamp64/apps/test2/htdocs
#3) I type in a web browser. http://localhost/test2 to access it as a user

I am sorry. What am I missing or not understanding?

**Comment by glensc** [comment]
you need to edit `routes.yml` file:
- https://github.com/eventum/eventum/issues/754#issuecomment-565804770

**Comment by JimFromAustin** [comment]
Sorry Glen, I thought if I installed to alias directory I did not have to edit routes.yml.

I have now edited routes.yml by prefixing all routes with /test3/htdocs

This gets it started but there is still a problem. ( I should say I am not certain which direction to read the Symphony Exception report, but I'm assuming oldest information is at the bottom and last execution attempt is at the top of the file.)

The Exception report shows the following with no error, everything is fine:
-----------------------------
Kernel::handleRequest() 
in C:\wamp64\apps\test3\htdocs\setup\index.php (line 43) 
--------------

A little while later, it throws the following exception:
--------------
ResourceNotFoundException 
No routes found for "/test3/setup/".
--------------

Not sure what may be going on, but when I edit the routes.xml file to this it has no impact, as if the routes.xml file was never read.  It still gives same error.
--------------
setup:
    path: /test3lkajsdflkjadfljk/htdocs/setup/
    methods: GET|POST
    controller: Eventum\Controller\Setup\SetupController::defaultAction
--------------

I am back at 3.8.7, fresh install if that has any impact.

**Comment by glensc** [comment]
If your alias is `/test3` instead of `/` you need to prefix with `/test3`, not `/test3/htdocs` all `path:` keys in `routes.yml`.

do read my previous comments:
- https://github.com/eventum/eventum/issues/754#issuecomment-565804770

**Comment by JimFromAustin** [comment]
Thanks for getting back Glen. (read everything above and tried several tests )   

Here is where I am at:

I started over with a new install:

1) c:\eventum  (all files are unpacked here)
2) Alias /eventum c:/eventum/htdocs
3) prefixed all routes with /eventum in  file C:\eventum\res\routes.yml 
4) enter http://server/eventum into browser address bar 
4) Browser is redirected to http://server/eventum/login.php

Browser indicates "The page isn’t redirecting properly"

Dont think it is a permission problem related to my network or Apache, because using the same Apache server I can setup a different Alias to an internal website and have no problem accessing it.

**Comment by JimFromAustin** [comment]
From what I can tell, it appears Symfony Guard authentication is failing.  (I know zero about Symfony)  I can see c:\eventum\res\packages\security.yml that guard authentication is enabled, but I would not know where to start on why it failed failed or how to fix it.

**Comment by glensc** [comment]
from this access log:

```
"GET /eventum/ HTTP/1.1" 302 -
"GET /eventum/setup/ HTTP/1.1" 302 318
"GET /eventum/login.php HTTP/1.1" 302 -
"GET /eventum/setup/ HTTP/1.1" 302 318
```

I can say that routing problem is fixed now.

And seems that Eventum detects that installation is not finished and redirects to `/setup`, and setup sees that installation is finished and redirects to `/login`, and login thinks installation is not finished and redirects to `/setup`, so it's an endless loop.

They both check whether `config/setup.php` is present and not empty:
- https://github.com/eventum/eventum/blob/v3.8.7/lib/eventum/class.setup.php#L248-L257
- https://github.com/eventum/eventum/blob/v3.8.7/htdocs/setup/index.php#L22
- https://github.com/eventum/eventum/blob/v3.8.7/init.php#L19

**Comment by glensc** [comment]
hmm, setup should redirect to `/`, not `/login`:

```
"GET /eventum/setup/ HTTP/1.1" 302 318
"GET /eventum/login.php HTTP/1.1" 302 -
```

so you get into some different scenario...

**Comment by glensc** [comment]
so, yes, you need to update paths also in `security.yml`:
- `security.firewalls.dev.pattern`: https://github.com/eventum/eventum/blob/v3.8.7/res/packages/security.yml#L14
- `security.firewalls.main.pattern`: https://github.com/eventum/eventum/blob/v3.8.7/res/packages/security.yml#L26
- :https://github.com/eventum/eventum/blob/v3.8.7/res/packages/security.yml#L53

**Comment by JimFromAustin** [comment]
Thanks Glen.  That worked!  Prefixed L14, L26, L53 (all 4) with /eventum and installation moved forward and created database tables.  I forgot to check create database the firsttime, and had to delete /config/setup.php before it would install again.

I verified all PHP Extensions are running using phpinfo()

However,  I am not out of the woods.  A new problem that is preventing progress:

Failed to start the session because headers have already been sent by "C:\eventum\lib\eventum\class.language.php" at line 159.

RuntimeException:
Failed to start the session because headers have already been sent by "C:\eventum\lib\eventum\class.language.php" at line 159.

  at C:\eventum\vendor\symfony\http-foundation\Session\Storage\NativeSessionStorage.php:147
  at Symfony\Component\HttpFoundation\Session\Storage\NativeSessionStorage->start()
     (C:\eventum\vendor\symfony\http-foundation\Session\Storage\NativeSessionStorage.php:320)
  at Symfony\Component\HttpFoundation\Session\Storage\NativeSessionStorage->getBag('attributes')
     (C:\eventum\vendor\symfony\http-foundation\Session\Session.php:244)
  at Symfony\Component\HttpFoundation\Session\Session->getBag('attributes')
     (C:\eventum\vendor\symfony\http-foundation\Session\Session.php:266)
  at Symfony\Component\HttpFoundation\Session\Session->getAttributeBag()
     (C:\eventum\vendor\symfony\http-foundation\Session\Session.php:60)
  at Symfony\Component\HttpFoundation\Session\Session->has('_security.last_error')
     (C:\eventum\vendor\symfony\security-http\Authentication\AuthenticationUtils.php:45)
  at Symfony\Component\Security\Http\Authentication\AuthenticationUtils->getLastAuthenticationError()
     (C:\eventum\src\Controller\SecurityController.php:32)
  at Eventum\Controller\SecurityController->login(object(Request), object(AuthenticationUtils))
     (C:\eventum\vendor\symfony\http-kernel\HttpKernel.php:146)
  at Symfony\Component\HttpKernel\HttpKernel->handleRaw(object(Request), 1)
     (C:\eventum\vendor\symfony\http-kernel\HttpKernel.php:68)
  at Symfony\Component\HttpKernel\HttpKernel->handle(object(Request), 1, true)
     (C:\eventum\vendor\symfony\http-kernel\Kernel.php:201)
  at Symfony\Component\HttpKernel\Kernel->handle(object(Request))
     (C:\eventum\src\Kernel.php:79)
  at Eventum\Kernel::handleRequest()
     (C:\eventum\htdocs\index.php:18)
  at require('C:\\eventum\\htdocs\\index.php')
     (C:\eventum\htdocs\login.php:14)

**Comment by glensc** [comment]
- https://github.com/eventum/eventum/blob/v3.8.7/lib/eventum/class.language.php#L159

your [setlocale](http://php.net/setlocale) call is outputting something, likely a warning. 

also, you should use production php.ini, i.e [display_errors](http://php.net/display_errors) should never be enabled unless you are developing an application. I don't have any windows machine, so do not have idea what the warning could be.

- https://github.com/php/php-src/blob/php-7.4.1/php.ini-production
- https://github.com/php/php-src/blob/php-7.4.1/php.ini-development

**Comment by JimFromAustin** [comment]
Thanks Glen - I turned off warnings and that fixed the issue.  This was just a test install to see if I could even get it working.  We have an installation that has been in use since 2007.  My next goal will be to see if I can upgrade this installation.   We are using this in a production environment and need to be stable. My end users would be clueless if anything goes wrong.  They are not software developers.  They are paper work administrators.  We have configured Eventum to track Office issues, not software issues.
